File: src/commands/loader.ts
import { existsSync, readdirSync, statSync } from "node:fs";
import { extname, join, relative } from "node:path";
import type { DiscoveryResult } from "../config/discovery.ts";
import { getHeddleHome, getLocalHeddleDir } from "../config/paths.ts";
import { appendMessage } from "../session/jsonl.ts";
import type { Message } from "../types.ts";
import type { SlashCommand } from "./types.ts";

function scanDirectory(dir: string, baseDir?: string): SlashCommand[] {
	if (!existsSync(dir)) return [];

	const commands: SlashCommand[] = [];
	const base = baseDir ?? dir;

	for (const entry of readdirSync(dir)) {
		const fullPath = join(dir, entry);
		const stat = statSync(fullPath);

		if (stat.isDirectory()) {
			commands.push(...scanDirectory(fullPath, base));
			continue;
		}

		if (extname(entry) !== ".md") continue;

		const relPath = relative(base, fullPath);
		const name = relPath.replace(/\.md$/, "").replace(/\//g, ":");

		const filePath = fullPath;
		commands.push({
			name,
			description: `Custom command: ${name}`,
			execute: async (args, ctx) => {
				const content = await Bun.file(filePath).text();
				const userContent = args ? `${content}\n\n${args}` : content;
				const msg: Message = { role: "user", content: userContent };
				ctx.messages.push(msg);
				await appendMessage(ctx.sessionFile, msg);
				console.log(`  [skill] ${name} injected`);
			},
		});
	}

	return commands;
}

/**
 * Load custom commands from discovery levels and legacy paths.
 * Scans both `skills/` and `commands/` subdirs for backward compatibility.
 * Discovery levels are scanned in priority order (deepest first),
 * with later entries overriding earlier ones for same-named commands.
 */
export async function loadCustomCommands(discovery?: DiscoveryResult): Promise<SlashCommand[]> {
	const commandMap = new Map<string, SlashCommand>();

	if (discovery) {
		// Discovery-based loading: iterate levels in reverse (shallowest first)
		// so that deeper/more-specific levels override shallower ones
		const reversedLevels = [...discovery.levels].reverse();
		for (const level of reversedLevels) {
			const baseDir = level.source === "agents" ? level.path : level.path;
			const subdirs = level.source === "agents" ? [baseDir] : [join(baseDir, "skills"), join(baseDir, "commands")];

			for (const dir of subdirs) {
				for (const cmd of scanDirectory(dir)) {
					commandMap.set(cmd.name, cmd);
				}
			}
		}
	} else {
		// Legacy fallback: hardcoded 4-dir scan for backward compatibility
		const heddleHome = getHeddleHome();
		const localDir = getLocalHeddleDir();

		const dirs = [
			join(heddleHome, "skills"),
			join(heddleHome, "commands"),
			join(localDir, "skills"),
			join(localDir, "commands"),
		];

		for (const dir of dirs) {
			for (const cmd of scanDirectory(dir)) {
				commandMap.set(cmd.name, cmd);
			}
		}
	}

	return [...commandMap.values()];
}

File: src/commands/loader.rs
//! Discover custom commands as `.md` files in skills/ and commands/ dirs.

use std::path::{Path, PathBuf};
use std::sync::Arc;

use chrono::Utc;

use super::types::{CommandContext, SlashCommand};
use crate::config::discovery::{DiscoveryResult, DiscoverySource};
use crate::config::paths::{get_heddle_home, get_local_heddle_dir};
use crate::session::jsonl::append_message;
use crate::types::{Message, UserMessage};

fn scan_directory(dir: &Path, base: &Path) -> Vec<SlashCommand> {
    let mut out = Vec::new();
    let entries = match std::fs::read_dir(dir) {
        Ok(e) => e,
        Err(_) => return out,
    };
    for entry in entries.flatten() {
        let path = entry.path();
        let ft = match entry.file_type() {
            Ok(t) => t,
            Err(_) => continue,
        };
        if ft.is_dir() {
            out.extend(scan_directory(&path, base));
            continue;
        }
        if path.extension().and_then(|e| e.to_str()) != Some("md") {
            continue;
        }
        let rel = path.strip_prefix(base).unwrap_or(&path);
        let name = rel
            .with_extension("")
            .to_string_lossy()
            .replace('\\', "/")
            .replace('/', ":");
        let file_path: PathBuf = path.clone();
        let name_for_msg = name.clone();
        out.push(SlashCommand {
            name,
            description: format!("Custom command: {}", name_for_msg),
            execute: Arc::new(move |args: &str, ctx: &mut CommandContext<'_>| {
                let file_path = file_path.clone();
                let name_for_msg = name_for_msg.clone();
                let args = args.to_string();
                Box::pin(async move {
                    let content = match tokio::fs::read_to_string(&file_path).await {
                        Ok(c) => c,
                        Err(_) => return None,
                    };
                    let user_content = if args.is_empty() {
                        content
                    } else {
                        format!("{content}\n\n{args}")
                    };
                    let msg = Message::User(UserMessage {
                        content: user_content,
                    });
                    ctx.messages.push(msg.clone());
                    let _ = append_message(&ctx.session_file, &msg);
                    println!("  [skill] {name_for_msg} injected");
                    None
                })
            }),
        });
    }
    out
}

pub fn load_custom_commands(discovery: Option<&DiscoveryResult>) -> Vec<SlashCommand> {
    let mut command_map = std::collections::HashMap::new();
    if let Some(disc) = discovery {
        let mut levels = disc.levels.clone();
        levels.reverse();
        for level in &levels {
            let subdirs: Vec<PathBuf> = if matches!(level.source, DiscoverySource::Agents) {
                vec![level.path.clone()]
            } else {
                vec![level.path.join("skills"), level.path.join("commands")]
            };
            for dir in subdirs {
                for cmd in scan_directory(&dir, &dir) {
                    command_map.insert(cmd.name.clone(), cmd);
                }
            }
        }
    } else {
        let heddle_home = get_heddle_home();
        let local_dir = get_local_heddle_dir();
        let dirs = [
            heddle_home.join("skills"),
            heddle_home.join("commands"),
            local_dir.join("skills"),
            local_dir.join("commands"),
        ];
        for dir in dirs {
            for cmd in scan_directory(&dir, &dir) {
                command_map.insert(cmd.name.clone(), cmd);
            }
        }
    }
    command_map.into_values().collect()
}

#[allow(dead_code)]
fn _unused_chrono() -> String {
    Utc::now().to_rfc3339()
}
